Make functions in Rewritable tailrec

davidegrohmann opened the pull request.

davidegrohmann committed:
Make topDown tailrec

davidegrohmann committed:
Make bottomUp tailrec

davidegrohmann committed:
Do not leak internal implementations from the Rewritable framework

Mats-SX commented on community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala:
```diff
@@ -119,53 +119,75 @@ object inSequence {
   }
 
-  def apply(rewriters: Rewriter*) = new InSequenceRewriter(rewriters)
+  def apply(rewriters: Rewriter*): Rewriter = new InSequenceRewriter(rewriters)
```
`override`

davidegrohmann replied:
On an object?

Mats-SX replied:
You're right; I'm wrong.

Mats-SX commented on community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala:
```diff
@@ -119,53 +119,75 @@ object inSequence {
-
-  class TopDownRewriter(rewriter: Rewriter) extends Rewriter {
+  private class TopDownRewriter(rewriter: Rewriter) extends Rewriter {
     def apply(that: AnyRef): AnyRef = {
```
`override`

Mats-SX commented on community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala:
```diff
@@ -119,53 +119,75 @@ object inSequence {
   }
 
-  def apply(rewriter: Rewriter) = new TopDownRewriter(rewriter)
+  def apply(rewriter: Rewriter): Rewriter = new TopDownRewriter(rewriter)
```
`override` EDIT: No, wait this one is on the `object`, perhaps?

Mats-SX commented on community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala:
```diff
@@ -119,53 +119,75 @@ object inSequence {
-
-  class BottomUpRewriter(val rewriter: Rewriter) extends Rewriter {
+  private class BottomUpRewriter(val rewriter: Rewriter) extends Rewriter {
     def apply(that: AnyRef): AnyRef = {
```
`override`

Mats-SX commented on community/cypher/frontend-3.0/src/main/scala/org/neo4j/cypher/internal/frontend/v3_0/Rewritable.scala:
```diff
@@ -119,53 +119,75 @@ object inSequence {
   }
 
-  def apply(rewriter: Rewriter) = new BottomUpRewriter(rewriter)
+  def apply(rewriter: Rewriter): Rewriter = new BottomUpRewriter(rewriter)
```
`override`

davidegrohmann committed:
Add tree visiting control to `bottomUp` and `topDown`

A predicate can be passed to those rewriter to indicate when stopping
visiting a subtree.

Alos `replacer` rewriter has been removed since it is subsumed by the other two.
